File: packages/py/src/aibvf/validate.py
"""Hand-rolled validator for AI BVF v1.0 portfolios. Zero dependencies."""
from __future__ import annotations
import re
from typing import Any

from .taxonomy import INDUSTRIES, FUNCTIONS, AI_TIERS

_PILLARS = ("strategic_alignment", "financial_return", "change_enablement", "governance_risk")
_BUCKETS = ("Agent-Proof", "Agent-Augmented", "Agent-Replaceable")
_CLASSIFICATIONS = ("Accelerate", "Fix", "Stop")
_COMPLIANCE = ("eu_ai_act", "dora", "csrd", "gdpr_ai")
_ID_RE = re.compile(r"^[a-z0-9-]{1,64}$")


def validate(doc: Any) -> dict:
    """Validate a portfolio document against AI BVF v1.0.

    Returns { "valid": bool, "errors": [{"path": str, "msg": str}, ...] }.
    """
    errors: list[dict] = []

    def push(path: str, msg: str) -> None:
        errors.append({"path": path, "msg": msg})

    if not isinstance(doc, dict):
        return {"valid": False, "errors": [{"path": "", "msg": "Document must be a JSON object."}]}

    if doc.get("bvf_version") != "1.0":
        push("bvf_version", 'Required. Must be the string "1.0".')

    org = doc.get("organization")
    if not isinstance(org, dict):
        push("organization", "Required object.")
    else:
        if not isinstance(org.get("name"), str) or not org["name"]:
            push("organization.name", "Required string.")
        if org.get("industry") not in INDUSTRIES:
            push("organization.industry", f"Must be one of: {', '.join(INDUSTRIES)}.")
        if "revenue_eur" in org and (not isinstance(org["revenue_eur"], (int, float)) or org["revenue_eur"] < 0):
            push("organization.revenue_eur", "Must be a non-negative number.")

    inits = doc.get("initiatives")
    if not isinstance(inits, list) or not inits:
        push("initiatives", "Required. Must be a non-empty array.")
    else:
        for i, init in enumerate(inits):
            base = f"initiatives[{i}]"
            if not isinstance(init, dict):
                push(base, "Must be an object.")
                continue
            iid = init.get("id")
            if not isinstance(iid, str) or not _ID_RE.match(iid):
                push(f"{base}.id", "Required. Lowercase letters, digits, hyphens only (max 64).")
            name = init.get("name")
            if not isinstance(name, str) or not name:
                push(f"{base}.name", "Required string.")
            elif len(name) > 240:
                push(f"{base}.name", "Exceeds 240 characters.")
            if init.get("function") not in FUNCTIONS:
                push(f"{base}.function", f"Must be one of: {', '.join(FUNCTIONS)}.")
            if init.get("ai_tier") not in AI_TIERS:
                push(f"{base}.ai_tier", f"Must be one of: {', '.join(AI_TIERS)}.")
            if "bucket" in init and init["bucket"] not in _BUCKETS:
                push(f"{base}.bucket", f"Must be one of: {', '.join(_BUCKETS)}.")

            scores = init.get("scores")
            if not isinstance(scores, dict):
                push(f"{base}.scores", "Required object with all four pillar scores.")
            else:
                for p in _PILLARS:
                    ps = scores.get(p)
                    if not isinstance(ps, dict):
                        push(f"{base}.scores.{p}", "Required pillar score object.")
                        continue
                    v = ps.get("value")
                    if not isinstance(v, (int, float)) or v < 0 or v > 100:
                        push(f"{base}.scores.{p}.value", "Required. Number in [0, 100].")
                    if "confidence" in ps:
                        c = ps["confidence"]
                        if not isinstance(c, (int, float)) or c < 0 or c > 100:
                            push(f"{base}.scores.{p}.confidence", "Optional. Number in [0, 100].")

            if "classification" in init and init["classification"] not in _CLASSIFICATIONS:
                push(f"{base}.classification", f"Must be one of: {', '.join(_CLASSIFICATIONS)}.")
            if "decision_confidence" in init:
                v = init["decision_confidence"]
                if not isinstance(v, (int, float)) or v < 0 or v > 100:
                    push(f"{base}.decision_confidence", "Must be a number in [0, 100].")
            if "compliance" in init:
                if not isinstance(init["compliance"], list):
                    push(f"{base}.compliance", "Must be an array.")
                else:
                    for j, c in enumerate(init["compliance"]):
                        if c not in _COMPLIANCE:
                            push(f"{base}.compliance[{j}]", f"Must be one of: {', '.join(_COMPLIANCE)}.")

    return {"valid": not errors, "errors": errors}

File: packages/js/src/validate.ts
import { INDUSTRIES, FUNCTIONS, AI_TIERS } from './taxonomy.js';
import type { ValidationError, ValidationResult } from './types.js';

const PILLARS = ['strategic_alignment','financial_return','change_enablement','governance_risk'] as const;
const BUCKETS = ['Agent-Proof','Agent-Augmented','Agent-Replaceable'];
const CLASSIFICATIONS = ['Accelerate','Fix','Stop'];
const COMPLIANCE = ['eu_ai_act','dora','csrd','gdpr_ai'];

function isObj(x: unknown): x is Record<string, unknown> {
  return typeof x === 'object' && x !== null && !Array.isArray(x);
}

/**
 * Validates a portfolio document against the AI BVF v1.0 schema.
 * Deterministic, no network, no dependencies.
 */
export function validate(doc: unknown): ValidationResult {
  const errors: ValidationError[] = [];
  const push = (path: string, msg: string) => errors.push({ path, msg });

  if (!isObj(doc)) return { valid: false, errors: [{ path: '', msg: 'Document must be a JSON object.' }] };

  if (doc.bvf_version !== '1.0') push('bvf_version', 'Required. Must be the string "1.0".');

  const org = doc.organization;
  if (!isObj(org)) push('organization', 'Required object.');
  else {
    if (typeof org.name !== 'string' || !org.name) push('organization.name', 'Required string.');
    if (typeof org.industry !== 'string' || !INDUSTRIES.includes(org.industry as typeof INDUSTRIES[number])) {
      push('organization.industry', `Must be one of: ${INDUSTRIES.join(', ')}.`);
    }
    if (org.revenue_eur !== undefined && (typeof org.revenue_eur !== 'number' || org.revenue_eur < 0)) {
      push('organization.revenue_eur', 'Must be a non-negative number.');
    }
  }

  if (!Array.isArray(doc.initiatives) || doc.initiatives.length === 0) {
    push('initiatives', 'Required. Must be a non-empty array.');
  } else {
    doc.initiatives.forEach((init, i) => {
      const base = `initiatives[${i}]`;
      if (!isObj(init)) { push(base, 'Must be an object.'); return; }
      if (typeof init.id !== 'string' || !/^[a-z0-9-]{1,64}$/.test(init.id)) push(`${base}.id`, 'Required. Lowercase letters, digits, hyphens only (max 64).');
      if (typeof init.name !== 'string' || !init.name) push(`${base}.name`, 'Required string.');
      else if (init.name.length > 240) push(`${base}.name`, 'Exceeds 240 characters.');
      if (typeof init.function !== 'string' || !FUNCTIONS.includes(init.function as typeof FUNCTIONS[number])) push(`${base}.function`, `Must be one of: ${FUNCTIONS.join(', ')}.`);
      if (typeof init.ai_tier !== 'string' || !AI_TIERS.includes(init.ai_tier as typeof AI_TIERS[number])) push(`${base}.ai_tier`, `Must be one of: ${AI_TIERS.join(', ')}.`);
      if (init.bucket !== undefined && !BUCKETS.includes(init.bucket as string)) push(`${base}.bucket`, `Must be one of: ${BUCKETS.join(', ')}.`);

      const scores = init.scores;
      if (!isObj(scores)) push(`${base}.scores`, 'Required object with all four pillar scores.');
      else {
        for (const p of PILLARS) {
          // Both shapes are valid: a bare number, or { value, confidence? }.
          // Field reports showed agents naturally write scores: { pillar: 68 };
          // rejecting that was rigidity, not rigour.
          const ps = scores[p];
          if (typeof ps === 'number') {
            if (ps < 0 || ps > 100) push(`${base}.scores.${p}`, 'Number in [0, 100].');
            continue;
          }
          if (!isObj(ps)) { push(`${base}.scores.${p}`, 'Required. A number in [0, 100], or an object with a value in [0, 100].'); continue; }
          const v = ps.value;
          if (typeof v !== 'number' || v < 0 || v > 100) push(`${base}.scores.${p}.value`, 'Required. Number in [0, 100].');
          if (ps.confidence !== undefined && (typeof ps.confidence !== 'number' || ps.confidence < 0 || ps.confidence > 100)) {
            push(`${base}.scores.${p}.confidence`, 'Optional. Number in [0, 100].');
          }
        }
      }

      if (init.classification !== undefined && !CLASSIFICATIONS.includes(init.classification as string)) push(`${base}.classification`, `Must be one of: ${CLASSIFICATIONS.join(', ')}.`);
      if (init.decision_confidence !== undefined) {
        const v = init.decision_confidence;
        if (typeof v !== 'number' || v < 0 || v > 100) push(`${base}.decision_confidence`, 'Must be a number in [0, 100].');
      }
      if (init.compliance !== undefined) {
        if (!Array.isArray(init.compliance)) push(`${base}.compliance`, 'Must be an array.');
        else init.compliance.forEach((c, j) => {
          if (!COMPLIANCE.includes(c as string)) push(`${base}.compliance[${j}]`, `Must be one of: ${COMPLIANCE.join(', ')}.`);
        });
      }
    });
  }

  return { valid: errors.length === 0, errors };
}
